lightning/src/util/transaction_utils.rs

```diff
@@ -28,49 +28,42 @@ pub fn sort_outputs<T, C : Fn(&T, &T) -> Ordering>(outputs: &mut Vec<(TxOut, T)>
 	});
 }
 
-fn get_dust_value(output_script: &Script) -> u64 {
-	//TODO: This belongs in rust-bitcoin (https://github.com/rust-bitcoin/rust-bitcoin/pull/566)
-	if output_script.is_op_return() {
-		0
-	} else if output_script.is_witness_program() {
-		294
-	} else {
-		546
-	}
-}
-
 /// Possibly adds a change output to the given transaction, always doing so if there are excess
 /// funds available beyond the requested feerate.
 /// Assumes at least one input will have a witness (ie spends a segwit output).
 /// Returns an Err(()) if the requested feerate cannot be met.
-pub(crate) fn maybe_add_change_output(tx: &mut Transaction, input_value: u64, witness_max_weight: usize, feerate_sat_per_1000_weight: u32, change_destination_script: Script) -> Result<(), ()> {
+/// Returns the expected maximum weight of the fully signed transaction on success.
+pub(crate) fn maybe_add_change_output(tx: &mut Transaction, input_value: u64, witness_max_weight: usize, feerate_sat_per_1000_weight: u32, change_destination_script: Script) -> Result<usize, ()> {
 	if input_value > MAX_VALUE_MSAT / 1000 { return Err(()); }
 
+	const WITNESS_FLAG_BYTES: i64 = 2;
+
 	let mut output_value = 0;
 	for output in tx.output.iter() {
 		output_value += output.value;
 		if output_value >= input_value { return Err(()); }
 	}
 
-	let dust_value = get_dust_value(&change_destination_script);
+	let dust_value = change_destination_script.dust_value();
 	let mut change_output = TxOut {
 		script_pubkey: change_destination_script,
 		value: 0,
 	};
 	let change_len = change_output.consensus_encode(&mut sink()).unwrap();
-	let mut weight_with_change: i64 = tx.get_weight() as i64 + 2 + witness_max_weight as i64 + change_len as i64 * 4;
+	let mut weight_with_change: i64 = tx.get_weight() as i64 + WITNESS_FLAG_BYTES + witness_max_weight as i64 + change_len as i64 * 4;
 	// Include any extra bytes required to push an extra output.
 	weight_with_change += (VarInt(tx.output.len() as u64 + 1).len() - VarInt(tx.output.len() as u64).len()) as i64 * 4;
 	// When calculating weight, add two for the flag bytes
 	let change_value: i64 = (input_value - output_value) as i64 - weight_with_change * feerate_sat_per_1000_weight as i64 / 1000;
-	if change_value >= dust_value as i64 {
+	if change_value >= dust_value.as_sat() as i64 {
 		change_output.value = change_value as u64;
 		tx.output.push(change_output);
+		Ok(weight_with_change as usize)
 	} else if (input_value - output_value) as i64 - (tx.get_weight() as i64 + 2 + witness_max_weight as i64) * feerate_sat_per_1000_weight as i64 / 1000 < 0 {
-		return Err(());
+		Err(())
+	} else {
+		Ok(tx.get_weight() + WITNESS_FLAG_BYTES as usize + witness_max_weight)
 	}
-
-	Ok(())
```

Comment:
I think this patch or similar would be more readable:
```
diff --git a/lightning/src/util/transaction_utils.rs b/lightning/src/util/transaction_utils.rs
index 45ee1054..c170a2ce 100644
--- a/lightning/src/util/transaction_utils.rs
+++ b/lightning/src/util/transaction_utils.rs
@@ -50,7 +50,8 @@ pub(crate) fn maybe_add_change_output(tx: &mut Transaction, input_value: u64, wi
                value: 0,
        };
        let change_len = change_output.consensus_encode(&mut sink()).unwrap();
-       let mut weight_with_change: i64 = tx.get_weight() as i64 + WITNESS_FLAG_BYTES + witness_max_weight as i64 + change_len as i64 * 4;
+       let weight: i64 = tx.get_weight() as i64 + WITNESS_FLAG_BYTES + witness_max_weight as i64;
+       let mut weight_with_change: i64 = weight + change_len as i64 * 4;
        // Include any extra bytes required to push an extra output.
        weight_with_change += (VarInt(tx.output.len() as u64 + 1).len() - VarInt(tx.output.len() as u64).len()) as i64 * 4;
        // When calculating weight, add two for the flag bytes
@@ -59,10 +60,10 @@ pub(crate) fn maybe_add_change_output(tx: &mut Transaction, input_value: u64, wi
                change_output.value = change_value as u64;
                tx.output.push(change_output);
                Ok(weight_with_change as usize)
-       } else if (input_value - output_value) as i64 - (tx.get_weight() as i64 + 2 + witness_max_weight as i64) * feerate_sat_per_1000_weight as i64 / 1000 < 0 {
+       } else if (input_value - output_value) as i64 - weight * feerate_sat_per_1000_weight as i64 / 1000 < 0 {
                Err(())
        } else {
-               Ok(tx.get_weight() + WITNESS_FLAG_BYTES as usize + witness_max_weight)
+               Ok(weight)
        }
 }
```

Reply:
Done, with starting_weight instead of weight.